File: tigerline/corridor.py
"""Score corridor builder.

Templates are stored in ``corridor_templates.yaml`` with the favorite as home.
At runtime we flip the tuples if the away side is the actual favorite, so the
caller always reads (home_goals, away_goals) — same orientation as ReviewResult.
"""

from __future__ import annotations

from tigerline.models import MatchInput, ScenarioType, ScoreCorridor
from tigerline.rules import CorridorTemplate, load_corridor_templates


def build_corridor(
    scenario: ScenarioType,
    match: MatchInput,
    templates: dict[ScenarioType, CorridorTemplate] | None = None,
) -> ScoreCorridor:
    """Return the score corridor for ``scenario``, oriented to the match.

    Templates are written favorite-as-home. We mirror to (away, home) when the
    match's favorite is the away side.
    """
    if templates is None:
        templates = load_corridor_templates()

    template = templates.get(scenario)
    if template is None:
        # Defensive: a new scenario added to ScenarioType without a template.
        # Don't crash a live analysis — degrade to "unclear_skip".
        template = templates["unclear_skip"]

    flip = match.market.favorite == "away"
    scores = [_orient(s, flip) for s in template.scores]
    primary = _orient(template.primary, flip)

    # De-duplicate while preserving order (mirroring can produce duplicates if
    # symmetric, e.g. (1,1)).
    seen: set[tuple[int, int]] = set()
    deduped: list[tuple[int, int]] = []
    for s in scores:
        if s in seen:
            continue
        seen.add(s)
        deduped.append(s)

    if primary not in deduped:
        deduped.insert(0, primary)

    return ScoreCorridor(scores=deduped, primary=primary)


def _orient(score: tuple[int, int], flip: bool) -> tuple[int, int]:
    return (score[1], score[0]) if flip else score


__all__ = ["build_corridor"]

File: tigerline-web/js/corridor.js
/* =====================================================================
 * corridor.js — score corridor templates. Faithful port of
 * tigerline/corridor.py + config/corridor_templates.yaml.
 * Templates are written favorite-as-home; flip when the away side is fav.
 * ===================================================================== */
"use strict";

export const CORRIDOR_TEMPLATES = {
  two_goal_landing: { primary: [2, 0], scores: [[2, 0], [3, 0], [3, 1], [2, 1], [4, 1]] },
  must_win_pressure: { primary: [2, 0], scores: [[1, 0], [2, 0], [2, 1], [3, 1]] },
  pressure_under: { primary: [1, 1], scores: [[0, 0], [1, 0], [1, 1], [0, 1]] },
  goal_market: { primary: [2, 1], scores: [[2, 1], [2, 2], [1, 2], [3, 1], [1, 3]] },
  narrow_win: { primary: [1, 0], scores: [[1, 0], [2, 1], [2, 0], [1, 1]] },
  rotation_trap: { primary: [1, 1], scores: [[1, 1], [0, 1], [1, 2]] },
  unclear_skip: { primary: [1, 1], scores: [[1, 1]] },
};

const orient = ([h, a], flip) => (flip ? [a, h] : [h, a]);
const key = ([h, a]) => `${h}-${a}`;

/**
 * @param {string} scenario
 * @param {import('./types.js').MatchInput} match
 */
export function buildCorridor(scenario, match) {
  const template = CORRIDOR_TEMPLATES[scenario] || CORRIDOR_TEMPLATES.unclear_skip;
  const flip = match.market.favorite === "away";
  const primary = orient(template.primary, flip);

  const seen = new Set();
  const scores = [];
  for (const s of template.scores) {
    const oriented = orient(s, flip);
    const k = key(oriented);
    if (seen.has(k)) continue;
    seen.add(k);
    scores.push(oriented);
  }
  if (!seen.has(key(primary))) scores.unshift(primary);

  return { scores, primary };
}
